### PR TITLE
feature(files): adds new API for handling file uploads

### LINKED ISSUE #7778: Make hook to validate file in file upload action
To keep it general we should pass in the filename, mime, "simpletype", size, and the container it will be stored in (so, e.g. you could check a quota).

### LINKED ISSUE #9876: get_uploaded_file() doesn't handle multiple file uploads
It assumes only one file is being uploaded.

### LINKED ISSUE #9934: ElggFile should use move_uploaded_file()
Currently we populate ElggFiles by reading the whole contents of the uploaded file with `get_uploaded_file()`. We should instead add a method like `fromUploadedFile($file)` which would use http://php.net/manual/en/function.move-uploaded-file.php internally. This function ensures the file actually came from an upload. This would also avoid reading a potentially large file into memory, particularly bad because PHP already has to read it in the POST body.

### PR DESCRIPTION
**Pull request by mrclay**
Adds UploadsService and public APIs to access and move uploaded files into ElggFile objects.

Fixes #7778
Fixes #9876
Fixes #9934

(meant to replace #9929)

### CODE REVIEW

File: engine/classes/Elgg/EntityIconService.php [hunk 1]
```diff
@@ -425,6 +425,7 @@ public function hasIcon(\ElggEntity $entity, $size, $type = 'icon') {
 	 * @param string $entity_subtype Entity subtype
 	 * @param string $type           The name of the icon. e.g., 'icon', 'cover_photo'
 	 * @return array
+	 * @throws InvalidParameterException
```
Comment (hypeJunction) [inline engine/classes/Elgg/EntityIconService.php]: I didn't notice ;)

File: engine/classes/Elgg/UploadService.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		if ($uploaded_files instanceof UploadedFile) {
+			$uploaded_files = array($uploaded_files);
+		}
+
+		$wrapped_files = [];
+		if (empty($uploaded_files)) {
+			return $wrapped_files;
+		}
+
+		foreach ($uploaded_files as $index => $uploaded_file) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/UploadService.php]: If you are throwing out `\Elgg\UploadedFile` all of this is no longer needed. Just return `$uploaded_files`

File: docs/guides/file-system.rst [hunk 3]
```diff
@@ -101,8 +101,78 @@ by using ``touch()``.
...
+		'label' => 'Select images to upload',
+	]);
+
+
+.. code-block:: php
+
+	// in your action
+	foreach (elgg_get_uploaded_files('upload') as $upload) {
+	    $file = new ElggFile();
+		$file->owner_guid = elgg_get_logged_in_user_guid();
```
Comment (hypeJunction) [inline docs/guides/file-system.rst]: Indents

File: engine/tests/phpunit/Elgg/UploadServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+			return true;
+		});
+
+		$this->assertTrue($file->acceptUploadedFile($uploaded_file));
+		$this->assertEquals(1, $upload_event_calls);
+		$this->assertEquals(3, $upload_hook_calls);
+		$this->assertFalse($file->exists());
+	}
+
+	public function testCanSaveUploadToElggFile() {
```
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/UploadServiceTest.php]: This test become a dupe with these changes

**Comment by hypeJunction** [comment]
LGTM

File: docs/guides/file-system.rst [hunk 5]
```diff
@@ -101,8 +101,78 @@ by using ``touch()``.
...
+	]);
+
+
+.. code-block:: php
+
+	// in your action
+	$uploaded_files = elgg_get_uploaded_files('upload');
+	$uploaded_file = array_shift($uploaded_files);
+	if ($uploaded_file && !$uploaded_file->isValid()) {
+		$error = elgg_get_friendly_upload_error($uploaded_file->getError());
```
Comment (mrclay) [inline docs/guides/file-system.rst]: Did we add this function?

File: docs/guides/file-system.rst [hunk 6]
```diff
@@ -101,8 +101,78 @@ by using ``touch()``.
...
+		'help' => 'Only jpeg, gif and png images are supported',
+	]);
+
+
+.. code-block:: php
+
+	// in your action
+	$uploaded_files = elgg_get_uploaded_files('upload');
+	$uploaded_file = array_shift($uploaded_files);
+	if ($uploaded_file && !$uploaded_file->isValid()) {
```
Comment (mrclay) [inline docs/guides/file-system.rst]: Should first handle case of no uploads.

**Comment by mrclay** [comment]
Is anything here going to introduce a big roadblock when we try to add chunked upload support? Nothing jumps out at me, but the biggest difference is obviously that you can't run a mime detect until all the chunks are in. I'm ok if we have to add new API to support that.

**Comment by mrclay** [comment]
I'm thinking ahead a little to https://github.com/Elgg/Elgg/issues/4712. But this LGTM for now.

File: docs/guides/file-system.rst (hunk 5 above)
Comment (hypeJunction) [inline docs/guides/file-system.rst]: It's been around for a while. I think Brett added it a few years ago.

File: docs/guides/file-system.rst (hunk 6 above)
Comment (hypeJunction) [inline docs/guides/file-system.rst]: Probably.

**Comment by hypeJunction** [comment]
I haven't tried implementing a chunked upload yet, so I am not sure. I don't see how this would interfere - I assume you would be using a POST request with HMAC in the headers and bytes in the POST body, so I don't think Symfony files would even be suitable.

File: docs/guides/file-system.rst (hunk 6 above)
Comment (hypeJunction) [inline docs/guides/file-system.rst]: I don't care for the docs on this. There are enough examples in core.

**Comment by hypeJunction** [comment]
Champagne for everyone! :wine_glass:

**Pull request merged**